### DISCUSSION TITLE
How to run parametrized tests for an API?

### DISCUSSION THREAD

**Discussion by dragondive** [usebruno/bruno#2313] [Q&A]
As an example, consider the GraphQL countries API. URL: `https://countries.trevorblades.com/`

In the Query tab, I have the following:

```
query get_country($country_code: ID!) {
    country(code: $country_code) {
    	name
    }
}
```

In the Variables tab, I have the following:

```
{
  "country_code": "IN"
}
```

If I'd want to parametrize the `country_code` to run multiple independent tests of the API, how could I do it?

I tried writing a test like this:

```
test("should get country name", function() {
  bru.setVar("country_code","US")
  expect(res.getBody().data.country.name).to.equal('United States')
})
```

But it doesn't work:

```
✘  should get country name
expected 'India' to equal 'United States'
```

What other approaches can I try to run the API with various different values?

**Comment by Aliaic** [comment]
In theory, this is possible, but I'm not sure how bru's nextRequest works. You can refer to it.

Add two environment variables:
```
index: 0
country_array: [ { "index": 0, "code": "IN", "name": "India" }, { "index": 1, "code": "US", "name": "United States" }, { "index": 2, "code": "UK", "name": "United Kingdom" } ]
```
Pre-script

```
var country_codes = JSON.parse(bru.getEnvVar("country_array"));
var array_length = country_codes.length;
var index = bru.getEnvVar("index")
if (index<array_length) {
  var country = country_codes[index];
  bru.setVar("country_code", country.code);
}else{ 
  bru.setVar("country_code", "IN") //Fault tolerance
}
```

test script

```
var country_codes = JSON.parse(bru.getEnvVar("country_array"));
var array_length = country_codes.length;
var index = parseInt(bru.getEnvVar("index"));


if (index<array_length) {
  var country = country_codes[index];
  bru.setVar("country_code", country.code);
  var expected_country_name = country.name;
  var actual_country_name = res.body.data.country.name;  
  
  test("Country name is correct", function() {
  expect(actual_country_name).to.equal(expected_country_name)
});

  index++;
  bru.setEnvVar("index",index);
  bru.setNextRequest("test gQL");// change to your request name
  
} else {
  bru.setNextRequest(null);
  bru.setEnvVar("index",0);
}
```

**Reply by dragondive** [reply]
Thank you for the reply. This was helpful for me to learn more about Bruno's testing. Unfortunately, it appears that `setNextRequest()` doesn't actually result in a "loop" of requests (or at least, I couldn't get that working after plenty of trial-and-error). I tried unraveling the bruno code a little bit to figure out how the request collection is run, but given my time constraints, I couldn't make sufficient headway with that.

Nonetheless, my hunch is this feature is not made available by design because Collection Run Iterations and CSV Upload of input data are paid features in the Golden Edition (which is fair).

**Reply by robert7** [reply]
update Bruno 1.23.1:
it seems a "request loop" using ``bru.setNextRequest('XY')`` seems to work nicely when the test is run via GUI
but is ignored when run from the command line :(
